content/blog/2025-12-16-loop-parallelization-proposal.md.md

```diff
@@ -0,0 +1,552 @@
+- Reduces total number of parallel regions
+- Improves cache locality by processing arrays together
+
+## Implementation 3: Fusion + Shared Parallel Region
+
+The third implementation creates a shared parallel region for all fused loops:
+
+```cpp
+// Instead of creating separate parallel regions:
+#pragma omp parallel for
+for (...) { /* loop 1 */ }
+#pragma omp parallel for
+for (...) { /* loop 2 */ }
+
+// Create one shared region:
+#pragma omp parallel
+{
+    #pragma omp for
+    for (...) { /* loop 1 */ }
+    #pragma omp for
+    for (...) { /* loop 2 */ }
+}
+```
+
+**Expected Benefits:**
+- Minimize thread creation overhead (single fork/join)
+- Keep threads alive between loop iterations
+- Better thread affinity and cache effects
+
+# Experimental Setup
+
+## Benchmarks
+
+We evaluated our implementations on two benchmark suites:
+
+### 1. Synthetic Test Benchmarks
+Three kernels designed to test different parallelization scenarios:
+- **Array Ops**: Independent array operations (highly parallelizable)
+- **Scale & Offset**: Strided array accesses
+- **Element-wise**: Simple element-wise computations
+
+### 2. PolyBench/C Suite
+30 real-world scientific computing benchmarks including:
+- Linear algebra (gemm, syrk, seidel-2d)
+- Stencil computations (jacobi, fdtd-2d)
+- Data mining (correlation, covariance)
+
+## Methodology
+
+For each implementation, we:
+- Compiled benchmarks with `-O2` optimization
+- Ran with thread counts: 2, 4, 8
+- Compared against **serial baseline** (original code compiled without any parallelization pass applied)
+- Ran 5-10 iterations and computed mean/std deviation
+- Measured execution time using benchmark-internal timers
+
+**Hardware:** Apple M3 (8-10 cores)
+**LLVM Version:** 18.1.8
+**Compiler Flags:** `-O2 -fopenmp`
+
+# Results and Analysis
+
+## Test Benchmark Results
+
+![Test Performance Comparison](figures/tests_performance_comparison.png)
+
+The synthetic benchmarks reveal surprising results (speedup relative to serial baseline):
+
+| Implementation | Array Ops (8T) | Scale Offset (8T) | Elementwise (8T) |
+|----------------|----------------|-------------------|------------------|
+| Serial Baseline| 1.00x          | 1.00x             | 1.00x            |
+| Original       | 0.98x          | 0.99x             | 1.01x            |
+| Fusion         | 0.96x          | 0.97x             | 0.98x            |
+| Fusion+Shared  | 0.95x          | 0.96x             | 0.99x            |
+
+**Key Observations:**
+1. **Minimal speedup across all implementations** (< 1.05x)
+2. **Fusion variants actually slower** than the original
+3. **Performance degrades with optimization complexity**
+
+## PolyBench Results
+
+![PolyBench Speedup Comparison](figures/polybench_speedup_comparison.png)
+
+### Parallelization Coverage
+
+Out of 30 PolyBench benchmarks:
+- **Original**: 6 benchmarks parallelized (20%)
+- **Fusion**: 6 benchmarks parallelized (20%)
```

Comment:
How many times did loop fusion occur? I assume that this 6 counts even programs where basic parallelization was triggered, even when there was no fusion.